File: collection/All/untitled/src/BinaryTrees/DepthAndHeightOfTree.java
package BinaryTrees;

public class DepthAndHeightOfTree {

    private Node root = null;

    int value;

    public DepthAndHeightOfTree(int value){
        this.value = value;
    }




    /**
     * Determining the height of the tree
     */
    public void calculateHeight(){

        // Navigate to the value which we want to calculate the height
        Node current = root;
        while (current!=null){
            if(current.value==value){
                break;
            }
            if (value<current.value){
                current=current.left;
            } else {
                current=current.right;
            }
        }

        // Calculate the height
        int result = recursivelyCalculateHeight(current);
        System.out.println("Result:-> "+result);
    }

    /**
     * Determining the depth of the tree
     */
    public void calculateDepth(){
        int height=0;
        Node current = root;
        int result = recursivelyCalculateDepth(current,height);
        System.out.println("Result:-> "+result);
    }


    private int recursivelyCalculateDepth(Node current, int height) {

        if(current==null){
            return height;
        }

        if(value==current.value){
            // Breaking condition
            height = height+1;
            return height;
        }

        if(value<current.value){
            // Traverse the left subtree
            height = height+1;
            height = recursivelyCalculateDepth(current.left,height);
        }

        if(value>current.value){
            // Traverse the right subtree
            height = height+1;
            height = recursivelyCalculateDepth(current.right,height);
        }

        return height;
    }


    private int recursivelyCalculateHeight(Node current) {

       if(current==null){
           // Breaking condition:-> When we reach leaf node return zero
           return 0;
       }

        return 1+Math.max(
                recursivelyCalculateHeight(current.left),
                recursivelyCalculateHeight(current.right));
    }


    /** ************************ HELPER FUNCTIONS ********************** **/
    /**
     * Inserting a Node
     */
    public void insert(int value) {

        Node current = root;

        if(current==null){
            // There are no elements inserted, so we perform the first insertion operation
            root = createNode(value);
        }else{
            Node nodeToInsert = createNode(value);
            // Iterate to the leaf node from the root where we need to perform insertion
            while (current!=null){
                // Either we insert to left of root or we insert to right of root
                if(value < current.value){
                    if(current.right==null){
                        current.left = nodeToInsert;
                        break;
                    }else{
                        current = current.left;
                    }
                }else{
                    if(current.right==null){
                        current.right = nodeToInsert;
                        break;
                    }else{
                        current = current.right;
                    }
                }
            }
        }

    }

    public Node printResult(){
        return root;
    }

    private Node createNode(int value) {
        return new Node(value);
    }

    private class Node {
        Node left;
        Node right;
        int value;
        public  Node(int value){
            this.value = value;
        }

        @Override
        public String toString() {
            return "Node=" + value;
        }
    }
    /** ************************ HELPER FUNCTIONS ********************** **/
}

File: collection/Trees/BinaryTree/BinarySearchTree/DepthAndHeightOfTree/DepthAndHeightOfTree.kt
class DepthAndHeightOfTree {

    // Root Node
    var root : Node? = null
    // Keeping track of size
    private var size = 0





    fun calculateDepthOfNode(value : Int): Int {

        var depth = 0
        var current = root

        // No elements are present
        if(size==0){
            depth = -1
        }

        // Just one element is present
        if(size==1 && value==current?.info){
            return 1;
        }

        var temp = 0
        while (current!=null){

            if(value<current.info){
                // Traverse to left
                current = current.leftNode
                // Increment the counter
                temp++
            }else if(value>current.info){
                // Traverse to right
                current = current.rightNode
                // Increment the counter
                temp++
            }else if(value==current.info){
                depth = temp
                break
            }

        }


        return depth
    }


    fun calculateHeightOfNode(value : Int): Int {

        var current = root

        while(current!=null){

            // Try to find the node
            if(value==current.info){
                // We found the element ---> Now calculate the height of this element
               break;
            }

            if(value<current.info){
                // Traverse --> to left of the tree
                current = current.leftNode
            }else{
                // Traverse --> to right of the tree
                current = current.rightNode
            }

        }


        return recursivelyCalculateHeight(current!!)
    }

    private fun recursivelyCalculateHeight(current : Node?): Int {

        if(current==null){
            // We reached the leaf of the tree
            return 0;
        }

        return 1 + Math.max(
            recursivelyCalculateHeight(current.leftNode),
            recursivelyCalculateHeight(current.rightNode)
        )

    }



    /** ********************* HELPER FUNCTIONS ********************* **/
    class Node(val info : Int) {
        var leftNode : Node? = null
        var rightNode : Node? = null
        override fun toString() = "Node -> $info"
    }

    /**
     * Printing the root node
     */
    fun printNode() = root

    /**
     * Creating a new node
     */
    fun createNode(info : Int) : Node = Node(info)

    /**
     * Inserting a new element
     */
    fun insert(info : Int) {

        val newNode = Node(info)

        if(size==0){
            // If the element we are inserting is the first element, Then it becomes the root node
            root = newNode
            size+=1
        }else{
            // At-least there is one element ----> Now determine if we can add to left or if we can add to right
            var current = root

            while(current!=null){

                if(current.info==info){
                    // Don't do anything
                }else if(info<current.info){
                    if(current.leftNode==null){
                        // We reached to leaf position
                        // Insert to the left of the tree
                        current.leftNode = newNode
                        size += 1
                        break
                    }else{
                        // Traverse to left of the current node
                        current = current.leftNode
                    }
                }else{
                    if(current.rightNode==null){
                        // We reached to leaf position
                        // Insert to the right of the tree
                        current.rightNode = newNode
                        size += 1
                        break
                    }else{
                        // Traverse to right of the current node
                        current = current.rightNode
                    }
                }

            }
        }



    }
    /** ********************* HELPER FUNCTIONS ********************* **/


}